internal/internal_search_attributes.go

```diff
@@ -0,0 +1,535 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/sdk/converter"
+	"go.temporal.io/sdk/log"
+)
+
+type (
+	// SearchAttributes represents a collection of typed search attributes
+	SearchAttributes struct {
+		untypedValue map[SearchAttributeKey]interface{}
+	}
+
+	// SearchAttributeUpdate represents a change to SearchAttributes
+	SearchAttributeUpdate func(*SearchAttributes)
+
+	// SearchAttributeKey represents a typed search attribute key.
+	SearchAttributeKey interface {
+		// GetName of the search attribute.
+		GetName() string
+		// GetValueType of the search attribute.
+		GetValueType() enumspb.IndexedValueType
+		// GetReflectType of the search attribute.
+		GetReflectType() reflect.Type
+	}
+
+	baseSearchAttributeKey struct {
+		name        string
+		valueType   enumspb.IndexedValueType
+		reflectType reflect.Type
+	}
+
+	// SearchAttributeKeyString represents a search attribute key for a text attribute type.
+	SearchAttributeKeyString struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyKeyword represents a search attribute key for a keyword attribute type.
+	SearchAttributeKeyKeyword struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyBool represents a search attribute key for a boolean attribute type.
+	SearchAttributeKeyBool struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyInt64 represents a search attribute key for a integer attribute type.
+	SearchAttributeKeyInt64 struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyFloat64 represents a search attribute key for a float attribute type.
+	SearchAttributeKeyFloat64 struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyTime represents a search attribute key for a date time attribute type.
+	SearchAttributeKeyTime struct {
+		baseSearchAttributeKey
+	}
+
+	// SearchAttributeKeyKeywordList represents a search attribute key for a list of keyword attribute type.
+	SearchAttributeKeyKeywordList struct {
+		baseSearchAttributeKey
+	}
+)
```

Comment:
Can we do a `SearchAttributeKeyRawPayload`? As I found while doing Omes stuff, being able to pass though payloads is very convenient in some cases. We can add a docstring to the effect of "this is for advanced use cases only"

Reply:
We can, though I'm not as big of a fan of us adding tools for ourselves only to save a `switch` or similar. @Quinn-With-Two-Ns - thoughts? (also worth discussing whether it's this PR or another)

Reply:
I don't think it's only useful for us - it's useful when doing something like copying data from one workflow to another, or enabling things like the use case we were talking about where someone might scrape a workflow's history and apply things in it to the current workflow.

Reply:
That is a general problem with the inability to work with raw values in Go. The goal here was to not be able to provide the incorrect type even if you wanted to. But I don't mind the "raw" key if @Quinn-With-Two-Ns also agrees. Also open to other ideas that doesn't put it on par with all other properly-typed keys.

Reply:
Hm to copy from one workflow to another can't a user do `SearchAttributes.Copy()` ? If the only use is out Omes DSL I wouldn't add "raw" keys. If we did want it it should be in all languages not just Go, so maybe we should move the discussion out of this PR?

Reply:
Well they may not necessarily be other SAs is the problem. It could be input to the workflow, from a signal, etc, which you're accepting as a Payload

Reply:
@Sushisource can we open a features issue to discuss this? I think it's not included in Chads original proposal for TS attributes https://github.com/temporalio/proposals/pull/75 and if we did do it I'd want it in all langs that make sense not just Go.